ENH: gb.is_monotonic_increasing #17015

No-Stream opened the pull request:
- [x] closes [#17015](https://github.com/pandas-dev/pandas/issues/17015#issuecomment-316238679)
- [x] tests added / passed: 
updated test_tab_completion to expect additional functions `is_monotonic_increasing` and `is_monotonic_decreasing` 
added (semi-redundant) tests of gb.is_monotonically_increasing() - strictly increasing/decreasing, non-strictly, w/ inf vals, w/ nan vals
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry
added at the end of the `Other Enhancements` section

This is my first PR to Pandas, so I don't doubt that I'll need to edit things.

No-Stream committed:
ENH: gb.is_monotonic_increasing, is_monotonic_decreasing  #17015

jreback commented:
tests go in ``pandas/tests/groupby/test_groupby.py``

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -130,7 +130,7 @@ Other Enhancements
 - :func:`pd.read_sas()` now recognizes much more of the most frequently used date (datetime) formats in SAS7BDAT files (:issue:`15871`).
 - :func:`DataFrame.items` and :func:`Series.items` is now present in both Python 2 and 3 and is lazy in all cases (:issue:`13918`, :issue:`17213`)
-
+- :func: groupby.is_monotonic_increasing and .is_monotonic_decreasing extend Series.is_monotonic_increasing to groups, returning whether each group is monotonically increasing or decreasing, respectively. (:issue:`17015`)
```
this won't work for ``:func:`` as you have it written, I think enough is to say ``.is_monotonic_increasing/decreasing`` is added to ``.groupby()``

No-Stream replied:
Thanks for this fix and the above test location, which now seems painfully obvious. Will patch up the PR and fix tomorrow.

No-Stream committed:
added tests for gb.is_monotonically_increasing()/decreasing

No-Stream commented:
Added basic tests, for strictly monotonically increasing/decreasing, non-strictly, groups with inf, groups with nan, and to check method result == .apply(lambda x: x.is_monotonic_increasing) (though the method is just syntactic sugar in this implementation). (Not sure how much detail should be in the tests of gb.is_monotonic_increasing vs. Series.is_monotonic_increasing.)

MaximilianR commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -3843,6 +3843,128 @@ def test_cummin_cummax(self):
+        df = pd.DataFrame(source_dict)
+        result = df.groupby(['B']).C.is_monotonic_decreasing()
+        expected = pd.Series(index=['a', 'b', 'c', 'd'],
+                             data=[False, False, False, False],
```
FYI some of these may be less verbose as pytest parameterized tests, though not a big impact

Thanks for the PR!

No-Stream replied:
Thanks for the feedback. I did consider parametrizing, but when I've done it before, it inflated the test count (i.e. if I parametrized using the same test cases for both is_monotonic_increasing() and is_monotonic_decreasing, I'd have added two tests, rather than one. Maybe there's a way around that.)

jreback replied:
actually we would *prefer* to parametrize, it reduces the amount of duplication in code. feel free.

jreback commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -3843,6 +3843,128 @@ def test_cummin_cummax(self):
 
+    def test_is_increasing_is_decreasing(self):
+        # GH 17015
+
```
we don't really need all this, as ``is_montonic`` and friends are already well tested in ``Index`` (where they are actually implemented), but no big deal.

No-Stream replied:
figured as much but wasn't sure whether to test here and risk duplication or not test here and have "untested" code.

jreback commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -3843,6 +3843,128 @@ def test_cummin_cummax(self):
+                             name='C')
+        expected.index.name = 'B'
+        tm.assert_series_equal(result, expected)
+        # Also check result equal to manually taking x.is_monotonic_increasing.
```
blank line

jreback commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -3843,6 +3843,128 @@ def test_cummin_cummax(self):
+                             name='C')
+        expected.index.name = 'B'
+        tm.assert_series_equal(result, expected)
+        # Also check result equal to manually taking x.is_monotonic_increasing.
```
same

jreback reviewed (changes requested):
can you also update api.rst for these methods (and I suspect others might be missing as well for GroupBy and sub-classes)

No-Stream committed:
parametrized tests for gb.is_monotonic_increasing/decreasing

No-Stream committed:
parametrized tests for gb.is_monotonic_increasing/decreasing

No-Stream commented:
@jreback I went ahead and parametrized the tests. Wanted to have the method itself as a param as I could then just use one test function to test both is_monotonic_increasing and decreasing, but I couldn't figure out a way to pass a method as a param. 

Also added these to SeriesGroupBy in api.rst. Not sure what else I should add. If I run  the `api_rst_coverage.py` script I found mentioned in the docs, I get a gigantic laundry list of `Class members (other than those beginning with '_') missing from api.rst`. Let me know if you have a suggestion for that.

jreback reviewed (changes requested):
can you rebase

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -130,7 +130,7 @@ Other Enhancements
 - `read_*` methods can now infer compression from non-string paths, such as ``pathlib.Path`` objects (:issue:`17206`).
 - :func:`pd.read_sas()` now recognizes much more of the most frequently used date (datetime) formats in SAS7BDAT files (:issue:`15871`).
 - :func:`DataFrame.items` and :func:`Series.items` is now present in both Python 2 and 3 and is lazy in all cases (:issue:`13918`, :issue:`17213`)
```
move to 0.21.1

No-Stream committed:
removed edits to whatsnew 0.21.0

No-Stream committed:
ENH: gb.is_monotonic_increasing #17015 fix rebase conflicts

No-Stream committed:
ENH: gb.is_monotonic_increasing #17015 fix rebase conflicts

No-Stream committed:
ENH: gb.is_monotonic_increasing #17015 fix rebase conflicts

No-Stream committed:
ENH: gb.is_monotonic_increasing #17015 fix rebase conflicts

No-Stream commented:
I rebased to try to integrate changes to pandas-dev/pandas, but it seems the commit history is profoundly screwed up now. (It does look to me like the upstream changes are now reflected, though.) Could someone suggest how to fix this?

jreback commented:
@AdamShamlian you need to build the extensions

``python setup.py build_ext --inplace``

you should do this whenever you pull (it has no effect if they are already built).

extensions are undergoing a lot of renovation lately.

jreback committed:
rebase and cleanup

jreback commented:
@No-Stream I rebased you. please have a look thru and make any additional corrections

jreback commented on doc/source/whatsnew/v0.21.1.txt:
```diff
@@ -22,7 +22,7 @@ Other Enhancements
 
 - :meth:`Timestamp.timestamp` is now available in Python 2.7. (:issue:`17329`)
--
+- :func: groupby.is_monotonic_increasing and .is_monotonic_decreasing extend Series.is_monotonic_increasing to groups, returning whether each group is monotonically increasing or decreasing, respectively. (:issue:`17015`)
```
can add ``:func:`` where needed here

jreback commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -3701,6 +3701,64 @@ def test_cummin_cummax(self):
+            'C': in_vals}
+        df = pd.DataFrame(source_dict)
+        result = df.groupby(['B']).C.is_monotonic_increasing()
+        expected = pd.Series(index=['a', 'b', 'c', 'd'],
```
you can write this as ``Index(list('abcd'), name='B')``

jreback reviewed (changes requested):
small changes. pls rebase. ping on green.

ghasemnaddaf committed:
DOC: add docstring for MultiIndex.fillna (#18018) (#18269)

ENH: gb.is_monotonic_increasing #17015 fix rebase conflicts

parametrized tests for gb.is_monotonic_increasing/decreasing

ENH: gb.is_monotonic_increasing, is_monotonic_decreasing  #17015

added tests for gb.is_monotonically_increasing()/decreasing

parametrized tests for gb.is_monotonic_increasing/decreasing

ENH: gb.is_monotonic_increasing #17015 fix rebase conflicts

ENH: gb.is_monotonic_increasing #17015 fix rebase conflicts

rebase and cleanup

simplified test format

fixed whatsnew to include method tags

No-Stream committed:
ENH: gb.is_monotonic_increasing #17015 fix merge/rebase conflicts

No-Stream committed:
ENH: gb.is_monotonic_increasing #17015 tests revised and passing

No-Stream commented:
@jreback I've rebased to upstream/master, done the two changes requested, and tests are passing. 

By the way, for the test in question `pd.Series(index=list('abcd', data=out_vals, name='B')` actually doesn't work because the `tm.assert_series_equal()` test fails on the index names being different, not the series names being different. So I just define the index with its name and then use it in the series construction in order to avoid defining `series.index.name` manually.

No-Stream commented:
@jreback any chance you can take a look? Commit was clean when submitted, now would just need to redo edits to a few files to edits to match up with upstream.

jreback commented:
you will need to merge master and push again

No-Stream committed:
ENH: gb.is_monotonic_increasing #17015 merged to current upstream/master

No-Stream commented:
@jreback re-merged and pushed changes

jorisvandenbossche commented on pandas/core/groupby.py:
```diff
@@ -1826,6 +1826,55 @@ def pipe(self, func, *args, **kwargs):
+        Returns whether each group is monotonically increasing.
+
+        Equivalent to ``.apply(lambda x: x.is_monotonic_increasing)``.
+
```
Can you add a `.. versionadded:: 0.22.0` line here (and in the other as well)?

I also think we should keep this for 0.22.0, as it is a new feature, 0.21.1 is for bug fixes.

No-Stream replied:
@jorisvandenbossche thanks for the feedback. I've gone ahead and fixed the whatsnew references and added the ``.. versionadded:: 0.22.0`` on my local branch.

I wanted to clarify the `_series_apply_whitelist` thing. It appears there's a long list of methods in `_common_apply_whitelist` - you'd want me to add the new methods there? And would there need to be any other changes?

jorisvandenbossche replied:
I think you need to add it to `_series_apply_whitelist`, as it is only a Series method, not a DataFrame method. If you add that, you can remove the `def is_monotonic_..` functions/

jorisvandenbossche reviewed (commented):
Is there a reason we don't use `_series_apply_whitelist` for this?

If you add `'is_monotonic_increasing', 'is_monotonic_decreasing'` to  `_series_apply_whitelist` (in groupby.py) instead of the new definitions, I think all the tests should pass as well.

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -1172,4 +1172,3 @@ Other
 ^^^^^
 - Bug where some inplace operators were not being wrapped and produced a copy when invoked (:issue:`12962`)
 - Bug in :func:`eval` where the ``inplace`` parameter was being incorrectly handled (:issue:`16732`)
```
if you can revert this file

jreback commented on doc/source/whatsnew/v0.21.1.txt:
```diff
@@ -23,6 +23,7 @@ Other Enhancements
 
 - :meth:`Timestamp.timestamp` is now available in Python 2.7. (:issue:`17329`)
 - :class:`Grouper` and :class:`TimeGrouper` now have a friendly repr output (:issue:`18203`).
+- :meth: groupby.is_monotonic_increasing and :meth: .is_monotonic_decreasing extend :meth: Series.is_monotonic_increasing to groups, returning whether each group is monotonically increasing or decreasing, respectively. (:issue:`17015`)
```
let's move to 0.22

jreback commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -2600,6 +2600,60 @@ def test_cummin_cummax(self):
+        # abs val increasing (F), non-strictly increasing (T)
+        ([1, 2, 5, 3, 2, 0, 4, 5, -6, 1, 1],
+         [True, False, False, True]),
+        # Test with inf vals
```
can you add a blank line before comments

jreback commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -2600,6 +2600,60 @@ def test_cummin_cummax(self):
+            'C': in_vals}
+        df = pd.DataFrame(source_dict)
+        result = df.groupby('B').C.is_monotonic_increasing()
+        expected = pd.Series(index=list('abcd'), data=out_vals, name='C')
```
you can construct the ``index=Index(list('abcd'), name='B')``

jreback reviewed (changes requested):
minor comments, otherwise lgtm.

jreback commented:
can you rebase and update according to comments

No-Stream committed:
ENH: gb.is_monotonic_increasing #17015 minor fixes for @jreback

No-Stream committed:
ENH: gb.is_monotonic_increasing #17015 rebase to master

No-Stream committed:
ENH: gb.is_monotonic_increasing #17015 fix changed whatsnew

jreback commented on doc/source/whatsnew/v0.22.0.txt:
```diff
@@ -138,6 +138,8 @@ Other Enhancements
 - :func:`Series` / :func:`DataFrame` tab completion also returns identifiers in the first level of a :func:`MultiIndex`. (:issue:`16326`)
 - :func:`read_excel()` has gained the ``nrows`` parameter (:issue:`16645`)
 - :func:``DataFrame.to_json`` and ``Series.to_json`` now accept an ``index`` argument which allows the user to exclude the index from the JSON output (:issue:`17394`)
+- :meth: groupby.is_monotonic_increasing and :meth: .is_monotonic_decreasing extend :meth: Series.is_monotonic_increasing to groups, returning whether each group is monotonically increasing or decreasing, respectively. (:issue:`17015`)
```
these need to be 

``:func:`SeriesGroupBy.is_monotonic_increasing` `` (and so on)

jreback reviewed (changes requested):
lgtm. small doc-change.

No-Stream committed:
ENH: gb.is_monotonic_increasing #17015 fix 22.0 whatsnew

jorisvandenbossche reviewed (changes requested):
Please check if it is possible to completely remove the `is_monotonic_increasing` function definitions by adding them to `_series_apply_whitelist`

No-Stream committed:
NH: gb.is_monotonic_increasing #17015 alternate version to remove function defs

No-Stream committed:
ENH: gb.is_monotonic_increasing #17015 rebase to master

No-Stream commented:
@jorisvandenbossche was able to implement it as you requested. This change does make the syntax a bit uglier. i.e. it's `series_groupby.is_monotonic_increasing` rather than `series_groupby.is_monotonic_increasing()`. That makes it inconsistent with other groupby methods, but it's working fine.

jreback committed:
Merge branch 'master' into PR_TOOL_MERGE_PR_17453

jreback committed:
fixup whatsnew

jreback commented:
@gfyoung can you rebase this

gfyoung committed:
Resolve v0.23.0.txt merge conflict

gfyoung committed:
Merge branch 'master' into is-monotonic-increasing-gb-method

gfyoung commented:
@jreback : Resolved the merge conflict, though I don't know if the CI will be all green.

jorisvandenbossche committed:
typo

jreback commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -323,6 +323,7 @@ Other Enhancements
 
 - ``IntervalIndex.astype`` now supports conversions between subtypes when passed an ``IntervalDtype`` (:issue:`19197`)
 - :class:`IntervalIndex` and its associated constructor methods (``from_arrays``, ``from_breaks``, ``from_tuples``) have gained a ``dtype`` parameter (:issue:`19262`)
+- :func: groupby.is_monotonic_increasing and :func: .is_monotonic_decreasing extend :func: Series.is_monotonic_increasing to groups, returning whether each group is monotonically increasing or decreasing, respectively. (:issue:`17015`)
```
these will not render correctly

gfyoung replied:
Done.

jreback reviewed (changes requested):
i am ok with this being an attribute
it is consistent with Series/Index

gfyoung committed:
Add ticks for function rendering

TomAugspurger commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -323,6 +323,7 @@ Other Enhancements
 
 - ``IntervalIndex.astype`` now supports conversions between subtypes when passed an ``IntervalDtype`` (:issue:`19197`)
 - :class:`IntervalIndex` and its associated constructor methods (``from_arrays``, ``from_breaks``, ``from_tuples``) have gained a ``dtype`` parameter (:issue:`19262`)
+- :func:`groupby.is_monotonic_increasing` and :func:`.is_monotonic_decreasing` extend `:func: Series.is_monotonic_increasing` to groups, returning whether each group is monotonically increasing or decreasing, respectively. (:issue:`17015`)
```
This should be

    :func:`SeriesGroupBy.is_monotonic_increasing`

I also don't think "extend" is the right word. Perhaps just

    Added :func:`SeriesGroupBy.is_monotonic_increasing` and :func:`SeriesGroupBy.is_monotonic_decreasing` (:issue:`17015`)

gfyoung replied:
Fair enough.  Fixed.

gfyoung committed:
Make whatsnew statement more concise

jreback commented:
look ok to me. @gfyoung can you rebase again.

gfyoung commented:
@jreback @jorisvandenbossche : CI is passing, and the PR is mergeable.  PTAL.

gfyoung commented:
@jreback @jorisvandenbossche : CI is passing, and the PR is mergeable.  PTAL.

jorisvandenbossche commented:
Thanks @No-Stream and @gfyoung
